File: src/mbs/application/article/rate.py
from typing import Protocol

from pydantic import BaseModel

from mbs.application.common.article_gateway import ArticleReader, ArticleRater
from mbs.application.common.exceptions import NotFound, Unauthorized, Forbidden
from mbs.application.common.id_provider import IdProvider
from mbs.application.common.interactor import Interactor
from mbs.domain.models import ArticleId
from mbs.domain.services.access import AccessService
import mbs.domain.exceptions as domain_exceptions


class ArticleGateway(ArticleReader, ArticleRater, Protocol):
    pass


class RateArticle(Interactor[ArticleId, None]):

    def __init__(
            self,
            article_gateway: ArticleGateway,
            access_service: AccessService,
            id_provider: IdProvider
    ):
        self._article_gateway = article_gateway
        self._access_service = access_service
        self._id_provider = id_provider

    async def __call__(self, data: ArticleId) -> None:
        try:
            self._access_service.ensure_can_rate_article(
                is_auth=self._id_provider.is_auth(),
                permissions=self._id_provider.permissions(),
                user_state=self._id_provider.user_state(),
            )
        except domain_exceptions.AccessDenied as error:
            raise Forbidden(str(error))
        except domain_exceptions.AuthenticationError as error:
            raise Unauthorized(str(error))

        article = await self._article_gateway.get_article(data)
        if not article:
            raise NotFound("Статья не найдена")

        await self._article_gateway.rate_article(data, self._id_provider.user_id())

File: src/application/article/rate.rs
use serde::Deserialize;
use utoipa::IntoParams;
use crate::application::common::article_gateway::ArticleGateway;
use crate::application::common::{
    error::AppError,
    id_provider::IdProvider,
    interactor::Interactor
};
use crate::domain::{
    models::{
        article::ArticleId,
        rate_state::RateState
    },
    services::access::ensure_can_rate_article
};

#[derive(Deserialize, IntoParams)]
pub struct RateArticleInput {
    #[param(example = uuid::Uuid::new_v4, value_type=uuid::Uuid)]
    pub id: ArticleId,
    pub state: RateState
}

pub struct RateArticle<'interactor> {
    pub id_provider: Box<dyn IdProvider>,
    pub article_gateway: &'interactor dyn ArticleGateway
} 

impl Interactor<RateArticleInput, ()> for RateArticle<'_> {
    async fn execute(&self, input: RateArticleInput) -> Result<(), AppError> {
        let article = match self.article_gateway.get_article(&input.id).await? {
            Some(article) => article,
            None => return Err(AppError::NotFound("id".into()))
        };

        ensure_can_rate_article(
            self.id_provider.permissions(),
            self.id_provider.user_state(),
            &article.state,
        )?;

        self.article_gateway.rate_article(&article.id, self.id_provider.user_id(), &input.state).await?;
        Ok(())
    }
}
